File: bindings/node/examples/oprf-client.mjs
// Cliente OPRF de referencia (Node) para un quipu-oprf-server.
//
// Flujo verificable: blind -> POST /v1/oprf/evaluate -> finalize (verifica la
// prueba DLEQ contra la clave pública fijada) -> secreto endurecido.
//
// La clave pública se FIJA fuera de banda. Pídesela al servidor UNA vez:
//
//   curl -s $QUIPU_OPRF_URL/v1/public-key
//
// y pásala como configuración. No se pide en cada llamada, a propósito: un
// servidor que te entrega la clave contra la que se le verifica no queda
// verificado en absoluto. La prueba DLEQ responde "¿es este el servidor que
// fijé?"; preguntárselo a él no es una respuesta.
//
//   QUIPU_OPRF_URL=https://oprf.tudominio.com \
//   QUIPU_OPRF_API_KEY=quipu_live_... \
//   QUIPU_OPRF_PUBKEY=<64 hex> \
//   node examples/oprf-client.mjs "mi-contraseña"

import { oprfHarden, OprfUnavailable, OprfRejected } from '../src/index.js';

const password = process.argv[2];
if (!password) {
  console.error('uso: node oprf-client.mjs <contraseña>');
  process.exit(1);
}

const baseUrl = process.env.QUIPU_OPRF_URL ?? 'http://127.0.0.1:8787';
const apiKey = process.env.QUIPU_OPRF_API_KEY;
if (!apiKey) {
  console.error('falta QUIPU_OPRF_API_KEY');
  process.exit(1);
}
const pinned = process.env.QUIPU_OPRF_PUBKEY;
if (!pinned) {
  console.error(`falta QUIPU_OPRF_PUBKEY (64 hex). Obtenla una vez con:
  curl -s ${baseUrl}/v1/public-key`);
  process.exit(1);
}

try {
  const secret = await oprfHarden({
    baseUrl,
    apiKey,
    password: Buffer.from(password, 'utf8'),
    serverPublicKey: Buffer.from(pinned.trim(), 'hex'),
  });
  console.log('secreto endurecido:', secret.toString('hex'));
} catch (e) {
  // Dos fallos distintos que exigen reacciones opuestas.
  if (e instanceof OprfRejected) {
    console.error('RECHAZADO: la prueba no valida contra la clave que fijaste.');
    console.error('No es la red. O el servidor rotó su clave, o no es el que crees.');
  } else if (e instanceof OprfUnavailable) {
    console.error('NO DISPONIBLE:', e.message);
    console.error('Reintentable. Nunca degrades a guardar la contraseña sin endurecer.');
  } else {
    throw e;
  }
  process.exit(1);
}

File: examples/oprf_client.py
"""Cliente OPRF de referencia (Python) para un `quipu-oprf-server`.

Flujo verificable de endurecimiento:
  1. quipu.voprf_blind(password)        -> el servidor nunca ve la contraseña.
  2. POST /v1/oprf/evaluate (API key)   -> evaluación + prueba DLEQ.
  3. quipu.voprf_finalize(...)          -> VERIFICA la prueba contra la clave
     pública FIJADA y deriva el secreto endurecido.

Uso:
  QUIPU_OPRF_URL=https://oprf.tudominio.com \\
  QUIPU_OPRF_API_KEY=quipu_live_... \\
  python examples/oprf_client.py "mi-contraseña"

En producción, FIJA (pin) la clave pública fuera de banda con QUIPU_OPRF_PUBKEY
en vez de pedirla al servidor.
"""

import json
import os
import sys
import urllib.request

import quipu


def _http(url, method, data=None, headers=None):
    req = urllib.request.Request(url, data=data, method=method, headers=headers or {})
    with urllib.request.urlopen(req) as r:
        return r.status, r.read().decode()


def harden(password, base_url, api_key, server_pub=None):
    """Devuelve el secreto endurecido (bytes) para `password`."""
    base = base_url.rstrip("/")
    if server_pub is None:
        _, body = _http(base + "/v1/public-key", "GET")
        server_pub = bytes.fromhex(json.loads(body)["public_key"])

    state, blinded = quipu.voprf_blind(password)
    _, body = _http(
        base + "/v1/oprf/evaluate",
        "POST",
        data=blinded.hex().encode(),
        headers={"Authorization": "Bearer " + api_key, "Content-Type": "text/plain"},
    )
    resp = json.loads(body)
    evaluated = bytes.fromhex(resp["evaluation"])
    proof = bytes.fromhex(resp["proof"])
    # Lanza ValueError si la prueba DLEQ no valida contra server_pub.
    return quipu.voprf_finalize(password, state, evaluated, proof, server_pub)


if __name__ == "__main__":
    if len(sys.argv) < 2:
        sys.exit("uso: oprf_client.py <contraseña>")
    url = os.environ.get("QUIPU_OPRF_URL", "http://127.0.0.1:8787")
    api_key = os.environ["QUIPU_OPRF_API_KEY"]
    pinned = os.environ.get("QUIPU_OPRF_PUBKEY")
    pub = bytes.fromhex(pinned) if pinned else None
    secret = harden(sys.argv[1].encode(), url, api_key, pub)
    print("secreto endurecido:", secret.hex())
